spring-core/src/main/java/org/togglz/spring/repository/ApplicationEventPublisherRepository.java

```diff
@@ -0,0 +1,43 @@
+
+package org.togglz.spring.repository;
+
+import org.springframework.context.ApplicationEventPublisher;
+import org.togglz.core.Feature;
+import org.togglz.core.repository.FeatureState;
+import org.togglz.core.repository.StateRepository;
+
+/**
+ * Implementation of {@link StateRepository} that publishes a {@link FeatureStateChangedEvent} when the
+ * {@link StateRepository#setFeatureState(FeatureState)} method is called
+ * 
+ * @author Igor Khudoshin
+ */
+public class ApplicationEventPublisherRepository implements StateRepository {
+
+    private final StateRepository delegate;
+
+    private ApplicationEventPublisher applicationEventPublisher;
+
+    /**
+     * Creates a facade for the supplied {@link StateRepository}.
+     * 
+     * @param delegate The repository to delegate invocations to
+     * @param applicationEventPublisher The {@link ApplicationEventPublisher} to publish the {@link FeatureStateChangedEvent} to
+     */
+    public ApplicationEventPublisherRepository(StateRepository delegate, ApplicationEventPublisher applicationEventPublisher) {
+        this.delegate = delegate;
+        this.applicationEventPublisher = applicationEventPublisher;
+    }
+
+    @Override
+    public FeatureState getFeatureState(Feature feature) {
+        return delegate.getFeatureState(feature);
+    }
+
+    @Override
+    public void setFeatureState(FeatureState featureState) {
+        FeatureState previousFeatureState = getFeatureState(featureState.getFeature());
+        delegate.setFeatureState(featureState);
+        applicationEventPublisher.publishEvent(new FeatureStateChangedEvent(previousFeatureState, featureState));

```

Comment:
Yeah, I guess this would make sense. However, there is currently no simple way to compare `FeatureState` instances. Maybe `FeatureState` should simply implement `equals` and `hashcode` correctly!?!

Reply:
I was looking into `FeatureState` and you're right. And even if we implement it, there's no way we can prevent cases like:

```java
FeatureState s = stateRepo.getFeatureState(FOO);
s.setEnabled(true);
stateRepo.setFeatureState(s);
```

In that case, unless stateRepo returns a copy, which is not the case for `InMemoryStateRepository`,  we won't be able to detect that change.

This leads me to another problem then: `previousFeatureState` probably won't reflect the previous state at all because of what I just said. Of course, we can prevent this from happening if we ensure we always copy the state when receiving or returning it from a state repository.

Reply:
Yeah, this is not easy to solve. Maybe we should keep it as it is for now and simply fire the event for all calls to `setFeatureState()`.

Reply:
We can do that. Nevertheless, this will expose that problem I described above, but this can be addressed in another PR, because it has more implications (it requires a review of state repositories).

Reply:
Yeah, I agree. We should address this problem later on. So I'll merge this now.